sim/FreeRTOS.cpp

```diff
@@ -1,3 +1,44 @@
 #include "FreeRTOS.h"
+#include <algorithm>
+#include <map>
+#include <stdio.h>
+#include <stdlib.h>
 
 void NVIC_SystemReset(void) {}
+
+void APP_ERROR_HANDLER(int err) {
+  fprintf(stderr, "APP_ERROR_HANDLER: %d", err);
+}
+
+namespace {
+std::map<void *, size_t> allocatedMemory;
+size_t currentFreeHeap = configTOTAL_HEAP_SIZE;
+size_t minimumEverFreeHeap = configTOTAL_HEAP_SIZE;
+}
+
+void *pvPortMalloc( size_t xWantedSize ) {
+  void* ptr = malloc(xWantedSize);
+  allocatedMemory[ptr] = xWantedSize;
+
+  size_t currentSize = 0;
+  std::for_each(allocatedMemory.begin(), allocatedMemory.end(), [&currentSize](const std::pair<void*, size_t>& item){
+    currentSize += item.second;
+  });
```

Comment:
Nice use of `<algorithm>`, but we can be even more explicit using `std::accumulate` from `<numeric>`

```suggestion
  const size_t currentSize = std::accumulate(
    allocatedMemory.begin(), allocatedMemory.end(), 0,
    [](const size_t lhs, const std::pair<void*, size_t>& item){
      return lhs + item.second;
  });
```

Reply:
Nice, it's even better with std::accumulate!